lightstep/sdk/metric/internal/syncstate/sync.go

```diff
@@ -178,74 +222,179 @@ func capture[N number.Any, Traits number.Traits[N]](_ context.Context, inst *Ins
 }
 
+func fingerprintAttributes(attrs []attribute.KeyValue) uint64 {
+	var fp uint64
+	for _, attr := range attrs {
+		fp += fprint.Mix(
+			fprint.FingerprintString(string(attr.Key)),
+			fingerprintValue(attr.Value),
+		)
+	}
+
+	return fp
+}
+
+func fingerprintSlice[T any](slice []T, f func(T) uint64) uint64 {
+	var fp uint64
+	for _, item := range slice {
+		fp += f(item)
+	}
+	return fp
+}
+
+func fingerprintValue(value attribute.Value) uint64 {
+	switch value.Type() {
+	case attribute.BOOL:
+		return fprint.FingerprintBool(value.AsBool())
+	case attribute.INT64:
+		return fprint.FingerprintInt64(value.AsInt64())
+	case attribute.FLOAT64:
+		return fprint.FingerprintFloat64(value.AsFloat64())
+	case attribute.STRING:
+		return fprint.FingerprintString(value.AsString())
+	case attribute.BOOLSLICE:
+		return fingerprintSlice(value.AsBoolSlice(), fprint.FingerprintBool)
+	case attribute.INT64SLICE:
+		return fingerprintSlice(value.AsInt64Slice(), fprint.FingerprintInt64)
+	case attribute.FLOAT64SLICE:
+		return fingerprintSlice(value.AsFloat64Slice(), fprint.FingerprintFloat64)
+	case attribute.STRINGSLICE:
+		return fingerprintSlice(value.AsStringSlice(), fprint.FingerprintString)
+	}
+
+	return 0
+}
+
+func sliceEqual[T comparable](a, b []T) bool {
+	if len(a) != len(b) {
+		return false
+	}
+	for i := range a {
+		if a[i] != b[i] {
+			return false
+		}
+	}
+	return true
+}
+
+// attributesEqual returns true if two slices are exactly equal.
+func attributesEqual(a, b []attribute.KeyValue) bool {
+	if len(a) != len(b) {
+		return false
+	}
+	for i := range a {
+		if a[i].Key != b[i].Key {
+			return false
+		}
+		switch a[i].Value.Type() {
+		case attribute.INVALID, attribute.BOOL, attribute.INT64, attribute.FLOAT64, attribute.STRING:
+			if a[i].Value != b[i].Value {
+				return false
+			}
+		case attribute.BOOLSLICE:
+			if !sliceEqual(a[i].Value.AsBoolSlice(), b[i].Value.AsBoolSlice()) {
+				return false
+			}
+		case attribute.INT64SLICE:
+			if !sliceEqual(a[i].Value.AsInt64Slice(), b[i].Value.AsInt64Slice()) {
+				return false
+			}
+		case attribute.FLOAT64SLICE:
+			if !sliceEqual(a[i].Value.AsFloat64Slice(), b[i].Value.AsFloat64Slice()) {
+				return false
+			}
+		case attribute.STRINGSLICE:
+			if !sliceEqual(a[i].Value.AsStringSlice(), b[i].Value.AsStringSlice()) {
+				return false
+			}
+		}
+
+	}
+	return true
+}
+
 // acquireRead acquires the read lock and searches for a `*record`.
-func acquireRead(inst *Instrument, aset attribute.Set) *record {
+func acquireRead(inst *Instrument, fp uint64, attrs []attribute.KeyValue) *record {
 	inst.lock.RLock()
 	defer inst.lock.RUnlock()
 
-	if rec, ok := inst.current[aset]; ok {
-		// Existing record case.
-		if rec.refMapped.ref() {
-			// At this moment it is guaranteed that the
-			// record is in the map and will not be removed.
-			return rec
-		}
+	rec := inst.current[fp]
+
+	// Note: we could (optionally) allow collisions and not scan this list.
```

Comment:
Good note for a future configuration option :)

Reply:
I wonder how much it actually matters, looking forward to real-world benchmarks.